### ISSUE TITLE
Update SBEMU to work with the the the port-trapping API in the upstream HX version

### ISSUE THREAD

**Issue by volkertb** [crazii/SBEMU#40]
@crazii To prevent yourself from having to maintain a fork of the HX project indefinitely, perhaps it's better to update SBEMU to be compatible with the port-trapping API that was introduced in more recent releases of the upstream HX/HDPMI project?

This is what I'm referring to: https://github.com/Baron-von-Riedesel/HX/issues/11#issuecomment-1773962231

Thanks.

**Comment by crazii** [comment]
I don't remember the details but SBEMU uses some extra functions that were added to the fork. It should be checked whether it could be removed easily.

**Comment by Torinde** [comment]
Would be great if those get in sync. 
DOSbox-X uses HX and I would like to use that together with SBEMU on a laptop with Intel HDA and booting DOS directly.

**Comment by crazii** [comment]
> Would be great if those get in sync. DOSbox-X uses HX and I would like to use that together with SBEMU on a laptop with Intel HDA and booting DOS directly.

HX has VSBHDA (https://github.com/Baron-von-Riedesel/VSBHDA), it's like SBEMU and works with latest HX, have you tried it?

**Comment by Torinde** [comment]
Will try. And now I see somebody tried already at Baron-von-Riedesel/VSBHDA/issues/2 (although with quite old version of DOSbox-X).

Still, it seems better if VSBHDA / SBEMU converge instead of diverge as currently.

**Comment by crazii** [comment]
> As answered [here](https://github.com/Baron-von-Riedesel/HX/issues/11#issuecomment-1875115515), the port-trapping API in the upstream HX project is documented at https://github.com/Baron-von-Riedesel/HX/blob/master/Src/HDPMI/HDPMIAPI.TXT.
> 
> So the question is if this provides all the needed functionality, or if some things are still unclear or missing.

> > > Or maybe we can help you document the API?
> > 
> > 
> > It has already been documented - see https://github.com/Baron-von-Riedesel/HX/blob/master/Src/HDPMI/HDPMIAPI.TXT
> 
> @crazii (SBEMU) and @bjt42 (SoftMPU), is this API and documentation sufficient for both of you to work with?

No. DPMI host with IOPL0 needs virtual interrupts, currently SetPVI is a unsafe hack that might cause problems. after digging into the freeze problem recent days, now I have to stick to the HDPMI fork with more code changes. see this comment for detail: https://github.com/crazii/SBEMU/issues/13#issuecomment-1938457427
Current HDPMI cannot work perfectly without huge changes/additions.

The idea of a new DPMI host comes to me recently that we solve this once and for all, and also for 16 bit protected mode games( ONE DPMI host can support 32bit and 16bit at the same time).
I could use the usbddos' 16bit DPMI code as code base, it is working with basic DPMI functions. it's hard (especially for compatibility tests/debugging) but very possible. By using C code we have some performance loss but more efficient in developing.  Besides the Open Watcom is a very optimized compiler set - far more optimized than Borland C last time I checked the assembly output.

if it is done, I can also create PRs for ADLiPT and SoftMPU so that they support PM games without big changes, even almost 0 changes/cost.

Sound intriguing, doesn't it? at least it does to me. :)

**Comment by crazii** [comment]
> you can make the new DPMI host somewhat compatible with HX?

It's definitely possible. I can try it.
I'm currently coding on it but there's more details in DPMI api than I thought, I'm not finishing the codes yet, and haven't started any tests/debugs yet.
The DPMI in usbddos implements only a small core portion of DPMI api, there's much more to do.
